pallets/parachain-staking/src/lib.rs

```diff
@@ -2021,6 +2021,76 @@ pub mod pallet {
 			(top_n, num_of_delegators, collator_stake, delegator_stake)
 		}
 
+		/// Sort the CandidatePool by total amount and return the best
+		/// `MaxSelectedCandidates` many candidates.
+		///
+		/// In case a collator from last round was replaced by a candidate with
+		/// the same total stake during sorting, we revert this swap to
+		/// prioritize collators over candidates.
+		///
+		/// # <weight>
+		/// Weight: O(N) where N is `MaxSelectedCandidates` bounded by
+		/// `MaxCollatorCandidates`
+		/// - Reads: CandidatePool, CollatorState, SelectedCandidates
+		/// # </weight>
+		fn get_collator_list(top_n: u32) -> Vec<T::AccountId> {
+			let mut candidates = <CandidatePool<T>>::get().into_vec();
+
+			log::trace!("{} Candidates for {} Collator seats", candidates.len(), top_n);
+
+			// Order candidates by their total stake (greatest to least)
+			candidates.sort_by(|a, b| b.amount.cmp(&a.amount));
+
+			// Choose the top MaxSelectedCandidates qualified candidates
+			let mut collators = candidates
+				.clone()
+				.into_iter()
+				.take(top_n as usize)
+				.filter(|x| x.amount >= T::MinCollatorStake::get())
+				.map(|x| x.owner)
+				.collect::<Vec<T::AccountId>>();
+
+			// Should never fail
+			let top_n = top_n.try_into().unwrap_or(usize::MAX);
+
+			// Check whether we wanted to replace a collator with a candidate which has
+			// equal stake, and if so, revert the swap.
+			//
+			// NOTE: Can only occur if we have filled all top_n slots and have a bigger pool
+			// than the number of collators.
+			if collators.len() < candidates.len() {
```

Comment:
The whole block is ugly. Would be happy about suggestions. Unfortunately all `get()`s return options and even though we should be safe to unwrap, we should never do that.

Reply:
We tried with Albi and could not find a better solution at the moment 😂